server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -134,95 +155,122 @@ public void afterRefresh(boolean didRefresh) {
-                                    localSegmentsPostRefresh,
-                                    storeDirectory,
-                                    indexShard.getOperationPrimaryTerm(),
-                                    segmentInfos.getGeneration()
-                                );
-                                localSegmentChecksumMap.keySet()
-                                    .stream()
-                                    .filter(file -> !localSegmentsPostRefresh.contains(file))
-                                    .collect(Collectors.toSet())
-                                    .forEach(localSegmentChecksumMap::remove);
-                                OnSuccessfulSegmentsSync();
-                                final long lastRefreshedCheckpoint = ((InternalEngine) indexShard.getEngine()).lastRefreshedCheckpoint();
-                                indexShard.getEngine().translogManager().setMinSeqNoToKeep(lastRefreshedCheckpoint + 1);
-                            } else {
-                                shouldRetry = true;
-                            }
+                String segmentInfoSnapshotFilename = null;
+                try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
+                    SegmentInfos segmentInfos = segmentInfosGatedCloseable.get();
+
+                    Collection<String> localSegmentsPostRefresh = segmentInfos.files(true);
+
+                    List<String> segmentInfosFiles = localSegmentsPostRefresh.stream()
+                        .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
+                        .collect(Collectors.toList());
+                    Optional<String> latestSegmentInfos = segmentInfosFiles.stream()
+                        .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
+
+                    if (latestSegmentInfos.isPresent()) {
+                        // SegmentInfosSnapshot is a snapshot of reader's view of segments and may not contain
+                        // all the segments from last commit if they are merged away but not yet committed.
+                        // Each metadata file in the remote segment store represents a commit and the following
+                        // statement keeps sure that each metadata will always contain all the segments from last commit + refreshed
+                        // segments.
+                        localSegmentsPostRefresh.addAll(SegmentInfos.readCommit(storeDirectory, latestSegmentInfos.get()).files(true));
+                        segmentInfosFiles.stream()
+                            .filter(file -> !file.equals(latestSegmentInfos.get()))
+                            .forEach(localSegmentsPostRefresh::remove);
+
+                        // Create a map of file name to size and update the refresh segment tracker
+                        Map<String, Long> currentLocalSizeMap = createSizeMap(localSegmentsPostRefresh);
```

Comment:
Do we need size map for all the segment files of a shard or just those files that are not yet uploaded?

Reply:
incorporated